pallets/migrations/src/lib.rs

```
@@ -0,0 +1,192 @@
+	fn migrate(&self, available_weight: Weight) -> Weight;
+}
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_std::prelude::*;
+
+	/// Pallet for migrations
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The list of migrations that will be performed
+		type MigrationsList: Get<Vec<Box<dyn Migration>>>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RuntimeUpgradeStarted(),
+		RuntimeUpgradeCompleted(Weight),
+		MigrationStarted(Vec<u8>),
+		MigrationCompleted(Vec<u8>, Weight),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// on_runtime_upgrade is expected to be called exactly once after a runtime upgrade.
+		/// We use this as a chance to flag that we are now in upgrade-mode and begin our
+		/// migrations.
+		fn on_runtime_upgrade() -> Weight {
+			log::warn!("Performing on_runtime_upgrade");
+
+			let mut weight: Weight = 0u64.into();
+			// TODO: derive a suitable value here, which is probably something < max_block
+			let available_weight: Weight = T::BlockWeights::get().max_block;
+
+			// start by flagging that we are not fully upgraded
+			<FullyUpgraded<T>>::put(false);
+			weight += T::DbWeight::get().writes(1);
+			Self::deposit_event(Event::RuntimeUpgradeStarted());
+
+			weight += perform_runtime_upgrades::<T>(available_weight.saturating_sub(weight));
+
+			if !<FullyUpgraded<T>>::get() {
+				log::error!(
+					"migrations weren't completed in on_runtime_upgrade(), but we're not
+				configured for multi-block migrations; state is potentially inconsistent!"
+				);
```

Should we panic in this case? If you panic the block is immediately invalid. But that may be better than getting one that we suspect is corrupted finalized in the relay chain.

---

I've generally made the assumption that if we fail during a runtime upgrade and can't produce a block that we'll just end up doing the same thing when we try to produce another block.

One reason to allow the block is that we are better off if we at least continue producing blocks (e.g. we could vote to modify storage, do another runtime upgrade, etc.)

I've definitely gone back and forth on this, though. It's a situation where we can't do anything good...